eye operator, for default storage type

ZiyueHuang opened the pull request:
## Description ##
eye operator, for default storage type

Requested in https://discuss.mxnet.io/t/is-there-an-eye-function-in-the-ndarray-api/526/5

cc @eric-haibin-lin 

## Checklist ##
### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] eye, unittest


## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

piiswrong commented on tests/python/unittest/test_ndarray.py:
```diff
@@ -736,6 +736,14 @@ def test_output():
     assert_almost_equal(arange_out.asnumpy(), np.arange(0, 20))
+    N_array = np.random.randint(1, high=3, size=3)
+    M_array = np.random.randint(1, high=3, size=3)
+    k_array = np.random.randint(-5, high=5, size=3)
```
Could you run some more trials?
At least locally & with larger ranges. Just to verify if there are edge cases

piiswrong replied:
Also I think there is the case where M is 0

cjolivier01 commented on src/operator/tensor/init_op.h:
```diff
@@ -63,6 +63,86 @@ struct InitOpParam : public dmlc::Parameter<InitOpParam> {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, DType* out_data, const nnvm::dim_t num_cols,
+                                  const nnvm::dim_t k) {
+    if ((i % num_cols) == ((i / num_cols) + k)) {
```
Looks like two divides per value-fill. Would it be faster ( at least for CPU), to fill with 0's and then "walk" across an offset (using only add) to fill in the one's?

piiswrong commented:
@ZiyueHuang ping

ZiyueHuang commented:
Thanks for your comments. Sorry for the late response.

@piiswrong I increased the size and the number of trials and added M=0 case in unittest. I also run 1000 times locally(ubuntu).

@cjolivier01 I changed to two kernels. About 10 times faster on cpu.

cjolivier01 commented on src/operator/tensor/init_op.h:
```diff
@@ -428,6 +405,39 @@ void FillComputeZerosEx(const nnvm::NodeAttrs& attrs,
+      if (nnz > 0) {
+        Kernel<eye_dns_fill<req_type>, xpu>::Launch(
+          s, nnz, out_data.dptr<DType>(),
+          std::max((nnvm::dim_t)0, param.k), param.k, num_cols);
```
nit: 
```cpp
static_cast<nnvm::dim_t>(0)
```
or
```cpp
0LL
```
although the latter sort of ties dim_t to a specific type.
